### ISSUE TITLE
recreate deletes all files using --patterns-from parameter

### ISSUE THREAD

**Issue by kcciti** [borgbackup/borg#4160]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Bug

## System information. For client/server mode post info for both machines.

Dell Laptop

#### Your borg version (borg -V).

borg 1.1.7

#### Operating system (distribution) and version.

Linux Solus v3.99999

#### Hardware / network configuration, and filesystems used.

ext4

#### How much data is handled by borg?

2.3M (test case)

#### Full borg commandline that lead to the problem (leave away excludes and passwords)
borg create --stat --verbose --list .::full ~/R_script
borg create --stat --verbose --list --patterns-from test.lst .::patterns
borg recreate --stat --verbose --list --patterns-from test.lst --target full.recreated .::full

CONTENTS: R_Script
R_Script/test.R
R_Script/.Rproj.user/7038483/pcs/files-pane.pper (and many other files under .Rproj.user)
R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[1].eps
R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[2].eps
R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[3].eps
R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[4].eps
R_Script/R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Post.eps

CONTENT:S test.lst
```
P re
R /home/user/R_script
+ .*Jags.*(1).*
- .*\.eps.*
- .*\.Rproj.*
```

## Describe the problem you're observing.

I run the full backup under with the full archive. The files look like they're all backed up under borg list .::full. Then, I create a new backup with --from-patterns test.lst. The backup seems to work as expected backing up everything but the *.eps and *.Rproj* files but keeps the R_Script/test.R and R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[1].eps file as specified by the 1st "+" line of test.lst. Finally, I run recreate on the full archive. I expected the results to be effectively same as the patterns backup in my full.recreated archive . However, it deletes everything but the R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[1].eps file. Even if I run the recreate command without --target and overwrite the full archive, R_Script/Jags-Ydich-XnomSsubj-MbernBeta-Diagtheta[1].eps is still the only file in the full archive after recreate.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Yes - please see steps above. I used a very small test directory.

#### Include any warning/errors/backtraces from the system logs

No warnings or errors.

**Comment by ThomasWaldmann** [comment]
Try absolute pathes.

If that does not resolve your problem, create a minimal, simple test case that reproduces the problem.

**Comment by kcciti** [comment]
ok, seems like it's working. For  others, the problem seems to be how I created the initial full backup. I didn't use the same exact path (ie. /home/user/R/) to backup but instead used ../R to backup and that caused issues when running through regex.

Thanks for the help.

**Comment by ThomasWaldmann** [comment]
You did a dry-run for the recreate?

**Comment by kcciti** [comment]
Sorry, it was full run - wasn't a dry run. I copied the wrong command. The dry-run outputs the exact actions with thte results as posted.

**Comment by ThomasWaldmann** [comment]
```
mkdir -p home/user/test
touch home/user/test/test_file3.txt
touch home/user/test/sijefliisje.txt
touch home/user/test/test_file1.txt
touch home/user/test/test_file2.txt

cat >test.lst <<EOF
P re
R home/user/test
+ .*test_file1.txt
- home/user/test/test.*
EOF

borg init -e none repo
borg create --verbose --list --stat repo::full home/user/test/
borg create --verbose --list --stat --patterns-from test.lst repo::patterns
borg recreate --verbose --stat --list --patterns-from test.lst --target full.recreated repo::full
```

this results in:
```
$ borg list repo::patterns
drwxrwxr-x tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/sijefliisje.txt
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/test_file1.txt

$ borg list repo::full.recreated
drwxrwxr-x tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/sijefliisje.txt
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/test_file1.txt

$ borg list repo::full
drwxrwxr-x tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/test_file3.txt
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/sijefliisje.txt
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/test_file1.txt
-rw-rw-r-- tw     tw            0 Tue, 2018-11-27 23:58:16 home/user/test/test_file2.txt
```

**Comment by ThomasWaldmann** [comment]
As you see, I used a slightly different test, using relative pathes instead of absolute and it worked.

I guess the problem you encountered is that when doing `borg create` using an absolute path pattern, it works because borg is *seeing* the full absolute path (in the filesystem).

OTOH, when using `borg recreate`, borg is processing pathes as they are stored in the archive. As you see above in the `repo::full` archive, these pathes are relative and never start with `/`, thus pattern matches trying to match that will all fail.

**Comment by kcciti** [comment]
I see - thanks

**Comment by ThomasWaldmann** [comment]
yeah, docs update is the least we need.

there is another ticket about absolute vs. relative, maybe this could be automatically done correctly by removing leading slashes when looking at path/file names in archive.

**Comment by fantasya-pbem** [comment]
> there is another ticket about absolute vs. relative
see #4029

**Comment by fantasya-pbem** [comment]
In #4029 we made a change to accept absolute paths, too. It's in the master branch and won't be released before 1.2.0.

**Comment by fantasya-pbem** [comment]
PR #5137 is for the 1.1 branch. For the master branch I plan to add a slightly different section.

**Comment by ThomasWaldmann** [comment]
OK, guess this is documented for 1.1 and documented and/or fixed for master (1.2).
